Adds ability to pass in map of file format options when creating a PailSpec.

dacamo76 opened the pull request:
By default the Pail file format is Hadoop SequenceFile.
There are three SequenceFileWriters each based on a compression type (none, record, block). The compression codec may also be specified.

This change allows us to pass compression type and compression codec options for Pail to use when creating the underlying SequenceFile.

dacamo76 committed:
Adds ability to pass in map of options when creating a PailSpec.

The options are passed on to the PailFormat.
By default the Pail file format is Hadoop SequenceFile and the available options are:

    1) Compression Type
        none, record, block
    2) Codec
        default, gzip, bzip2

dacamo76 committed:
Fixes error where PailSpec args are ignored.

Explicitly adds the file format name when creating the PailSpec.

When a Pail is created, the PailFormatFactory checks to see if there is
a PailSpec arg. If the arg is null or if the name of the PailSpec is
null then a new default PailSpec is created and used.

This necesitates a PailSpec name being explicitly set when options are
present, else a default PailSpec is used and the options are lost.

dcuddeback commented on src/main/clojure/clj_pail/core.clj:
```diff
@@ -1,11 +1,13 @@
+  ([^PailStructure structure]
+     (PailSpec. structure))
+  ([^PailStructure structure opts]
+     (PailSpec. PailFormatFactory/SEQUENCE_FILE opts structure)))
```
If I'm reading this correctly, this is going to use the following constructor:

``` java
public PailSpec(String name, Map<String, Object> args, PailStructure structure)
```

Which means the `PailSpec` will have a name of `"SequenceFile"`. Am I reading that correctly?

dacamo76 replied:
Correct.

dacamo76 replied:
Not sure if you're doubting the explicit argument `PailFormatFactory/SEQUENCE_FILE` in the call to `PailSpec`. I'm not a fan of explicitly setting arguments inside a function, but I think it's better than forcing a user to pass in a name to the `spec` function. My thinking is:
- This is the way `PailSpec` does it. If you don't pass in a name, it defaults to `"SequenceFile"`. Those familiar with Pail will be used to this behavior.
- At the moment, `"SequenceFile"` is the only pre-defined format.
- Forcing user to pass a name into `spec` adds complexity without much gain, most users will probably end up partially applying `spec` every time they use the library. 
- Passing in a name to `PailSpec` is the only way to get `PailSpec` to recognize the options. (commit message 862ab2c expands on this a little)
- The name must be `"SequenceFile"` else `PailSpec` ignores the options completely (nathanmarz/dfs-datastores#33). A name other than `"SequenceFile"` implies a custom implementation of `PailFormat`.

In my opinion, the correct way for `PailSpec` to handle this is to check for name and options parameters explicitly, if the name is empty and there are options present, use the default name `"SequenceFile"` and copy the options into the `PailSpec`. Unfortunately, Pail seems to be in sleep mode until the release of the Big Data book.

dcuddeback replied:
@dacamo76 Thanks for explaining your thought process on this. That makes a lot more sense now. I was confused by the purpose of a PailSpec's name. I'll take another look at this soon.

dacamo76 committed:
Adds file format parameter to PailSpec creation.

In a PailSpec, the options are tied specifically to the PailFormat used. The PailFormat is the parameter called "name" in the PailSpec constructors.

If using the options parameter, we need to know the PailFormat
used. Not passing in a PailFormat and letting the constructor set a
default format is dangerous, relying on implementation details.
We should always force a PailFormat name to be passed in when a PailSpec is created.

dacamo76 commented:
@dcuddeback After taking another look at this, I've changed course. We should force a user to pass in a "name" when using the options parameter. By implicitly setting the name to "SequenceFile" we are dangerously relying on implementations details.

The PailFormat is the parameter called "name" in the PailSpec constructors. The "options" are tied specifically to the "name" parameter. When you create a Pail, internally PailFormatFactory.create(spec) is called which instantiates a new PailFormat according to the "name" parameter in the spec, passing in the spec options as constructor args.

In other words, the PailSpec "name" parameter is used to instantiate a PailFormat of that name with the "options" parameter as its constructor arg.

dcuddeback commented:
@dacamo76 That makes sense to me, especially after reading `PailFormatFactory.create(spec)`.
